### ISSUE TITLE
Inconsistent computations when using different batch sizes for `lax.conv_general_dilated`

### ISSUE THREAD

**Issue by mohamad-amin** [google/jax#9102]
I observed some discrepancy when using a [Flax](https://github.com/google/flax) network to evaluate the network on some data in batch and loop modes, and I opened [an issue](https://github.com/google/flax/issues/1755) in the Flax repo.  

But seems like the discrepancy is caused by JAX at the first place:

```python
import numpy as np
from jax import lax, random

rng1, rng2 = random.split(random.PRNGKey(0))

lhs = random.normal(rng1, shape=((2, 1, 28, 28)))
rhs = random.normal(rng2, shape=((32, 1, 3, 3)))

logits = lax.conv_general_dilated(
    lhs=lhs,
    rhs=rhs,
    window_strides=(1, 1),
    padding="SAME")

logit = lax.conv_general_dilated(
    lhs=lhs[0:1],
    rhs=rhs,
    window_strides=(1, 1),
    padding="SAME")

print(np.allclose(logits, logit))  # outputs False
```

I was curious about why this happens, and if there is anyway that this can be fixed (generally, or a as a work-around, as my research is dependent on solving this problem right now), as it introduces some problems that are mentioned in [this](https://github.com/google/flax/issues/1755) issue.

P.S:
I thought about different order of sums, but I checked PyTorch's ResNet18 and the difference when I was evaluating a 128 batch at once and in loop was from order of 2.5e-11. There still is some difference, but it's 1e5 times smaller, which is a huge number in my opinion. Is there any difference between Jax's CUDA usage and Torch's CUDA usage? And is there any work-around regarding this use case?

**Comment by hawkinsp** [comment]
I think your reproduction is wrong. I ran:
```
import numpy as np
from jax import lax, random

rng1, rng2 = random.split(random.PRNGKey(0))

lhs = random.normal(rng1, shape=((2, 1, 28, 28)))
rhs = random.normal(rng2, shape=((32, 1, 3, 3)))

logits = lax.conv_general_dilated(
    lhs=lhs,
    rhs=rhs,
    window_strides=(1, 1),
    padding="SAME")

logit = lax.conv_general_dilated(
    lhs=lhs[0:1],
    rhs=rhs,
    window_strides=(1, 1),
    padding="SAME")

print(np.allclose(logits[0:1], logit))  # outputs False
```

(Note the extra `[0:1]` to get matching shapes) on a GPU colab and it succeded.

**Comment by mohamad-amin** [comment]
Thanks for the correction, sorry I copied the code without running, my bad.

**Comment by mohamad-amin** [comment]
Sorry, I coped the code without running it, my bad. Here's the correct reproduction:
```
import numpy as np
from jax import lax, random

rng1, rng2 = random.split(random.PRNGKey(0))

lhs = random.normal(rng1, shape=((128, 1, 28, 28)))
rhs = random.normal(rng2, shape=((32, 1, 3, 3)))

logits = lax.conv_general_dilated(
    lhs=lhs,
    rhs=rhs,
    window_strides=(1, 1),
    padding="SAME")

logits_loop = np.zeros_like(logits)
for i in range(128):
  logits_loop[i] = lax.conv_general_dilated(
    lhs=lhs[i:i+1],
    rhs=rhs,
    window_strides=(1, 1),
    padding="SAME")

print(np.allclose(logits[0:1], logit))
# Outputs False
```

**Comment by mohamad-amin** [comment]
Sorry, I copied the code without running it, my bad. Here's the correct reproduction:

```python
import numpy as np
from jax import lax, random

rng1, rng2 = random.split(random.PRNGKey(0))

lhs = random.normal(rng1, shape=((128, 1, 28, 28)))
rhs = random.normal(rng2, shape=((32, 1, 3, 3)))

logits = lax.conv_general_dilated(
    lhs=lhs,
    rhs=rhs,
    window_strides=(1, 1),
    padding="SAME")

logits_loop = np.zeros_like(logits)
for i in range(128):
  logits_loop[i] = lax.conv_general_dilated(
    lhs=lhs[i:i+1],
    rhs=rhs,
    window_strides=(1, 1),
    padding="SAME")

print(np.allclose(logits_loop, logit))
# Outputs False
```

**Comment by hawkinsp** [comment]
That reproduction is also buggy I think. Did you mean `logits` in the last line? It returns `True` for me with that change.

**Comment by marcvanzee** [comment]
Hi, sorry this is my fault. I made a mistake in the code and asked @mohamad-amin to copy it here. It is not clear the problem is actually in `conv_general_dilated`, so I will investigate it further in the Flax issue, which I will re-open now. Apologies!

**Comment by hawkinsp** [comment]
Closing since it sounds like we will look into this in the Flax issue.
